### ISSUE TITLE
[RFC] New Input component

### ISSUE THREAD

**Issue by bmdalex** [microsoft/fluent-ui-react#314]
# New Input component

## Problem description

The current `Input` component looks like this:

##### API
```jsx
<Input value="123" icon="search" placeholder="Search..." />
```

##### DOM structure:
```html
<div class="ui-input">
  <input type="text" placeholder="Search..." value="123" />
  <span class="ui-icon" aria-hidden="true"></span>
</div>
```

##### Rendered component:
![screen shot 2018-10-04 at 17 08 37](https://user-images.githubusercontent.com/5442794/46483534-32a08e80-c7f8-11e8-940c-80d9799ca6fe.png)

This introduces a few issues. Since it **directly** creates an HTML `<input />` element (implementation detail), it does not support a set of user scenarios:

#### 1. Styling the `<input />`
It's impossible to target the `<input />` element in order style it (e.g.: change the height of the input through styles) because it's not implemented as a shorthand slot.
Also, custom `styles` passed as prop will be set for the wrapping `<div />` instead of `<input />`:
```jsx
<Input placeholder="Search..." styles={{ backgroundColor: 'red' height: '80px' }} />
```
renders:
![screen shot 2018-10-04 at 17 24 53](https://user-images.githubusercontent.com/5442794/46484654-76949300-c7fa-11e8-872f-654dfb1944c9.png)
instead of expected:
![screen shot 2018-10-04 at 17 37 53](https://user-images.githubusercontent.com/5442794/46485595-41894000-c7fc-11e8-8b4c-f5b581b4c7cf.png)

#### 2. Setting HTML attributes for the `<input />`
It's impossible to target the `<input />` element in order to set attributes for it (e.g.: setting `tabIndex="-1"` to the `<input />` element so that it's skipped by keyboard navigation).
- passing `tabIndex="-1"` as prop adds it to the wrapping `<div />` so it doesn't work:
```jsx
<Input placeholder="Search..." tabIndex="-1" />
```
generates
```html
<div class="ui-input" tabindex="-1">
  <input type="text" placeholder="Search..." value="" />
</div>
```
- passing `input={{ tabIndex:"-1" }}` as prop does not target the `<input />` element because it's not implemented as a shorthand slot:
```jsx
<Input placeholder="Search..." input={{ tabIndex:"-1" }} />
```
generates
```html
<div class="ui-input" input="[object Object]">
  <input type="text" placeholder="Search..." value="" />
</div>
```
instead of expected
```html
<div class="ui-input">
  <input type="text" placeholder="Search..." value="" tabindex="-1" />
</div>
```

#### 3. Setting HTML attributes for the wrapping `<div />`
Because of implementation details, some HTML attributes (the ones specific to `<input />`) end up in the `<input />` element instead of the wrapping `<div />`:

```jsx
<Input placeholder="Search..." role="presentation" />
```
generates
```html
<div class="ui-input">
  <input type="text" placeholder="Search..." role="presentation" value="" />
</div>
```
instead of expected
```html
<div class="ui-input" role="presentation">
  <input type="text" placeholder="Search..." value="" />
</div>
```

## Proposed solutions

Both solutions rely on 2 main concepts:

#### i. Creating an additional standalone `InputBase` component

`InputBase` component will have the following structure:
##### Props:
```typescript
interface IInputBaseProps {
  as?: any // TODO: decide if we want it
  className?: string
  defaultValue?: React.ReactText
  fluid?: boolean
  inline?: boolean
  onChange?: ComponentEventHandler<IInputProps>
  placeholder?: string
  value?: React.ReactText
  type?: string
  styles?: IComponentPartStylesInput
  variables?: ComponentVariablesInput
}
```

##### API
```jsx
<Input placeholder="Search..." />
````

##### DOM structure
```html
<input type="text" placeholder="Search..." value="" />
```

For simple scenarios where other actions / elements are not needed, `InputBase` is all the user needs. It will also fix all scenarios mentioned in the `Problem description`:

##### 1. Styling the `<input />`
```jsx
<InputBase placeholder="Search..." styles={{ backgroundColor: 'red' height: '80px' }} />
```

##### 2. Setting HTML attributes for the `<input />`
```jsx
<InputBase placeholder="Search..." tabIndex="-1"  />
```

#### ii. Reimplementing `Input` using `InputBase` as shorthand
For scenarios like:
- input with `search` icon;
- clearable input;
- clearable input with `search` icon;

### Solution 1: `wrapper` prop

`Input` will leverage `InputBase` and use it internally.
All the `props` sent as `<Input ...{props} />` will be passed to the `InputBase` unless the user clearly specifies they are for the wrapping element using the `wrapper` prop.

##### Props:
```typescript
interface IInputProps extends IInputBaseProps {
  clearable?: boolean
  icon?: ItemShorthand
  wrapper?: ItemShorthand
}
```

##### API
```jsx
<Input {...inputProps} wrapper={wrapperProps} />
````
e.g.: 
```jsx
<Input
  placeholder="Search..."
  icon="search"
  clearable
  wrapper={{ role: "presentation" }}
/>
````

##### Component structure
```html
<ElementType ...> // <div />
  <InputBase ... />  // <input />
  <Icon ...>  // <span />
</ElementType>
```

##### DOM structure
```html
<div class="ui-input" role="presentation">
  <input type="text" placeholder="Search..." value="" />
  <span class="ui-icon" aria-hidden="true"></span>
</div>
```

This will solve all the scenarios mentioned above in the following fashion:

##### 1. Styling the `<input />`
```jsx
<Input
  placeholder="Search..."
  icon="search"
  clearable
  styles={{ backgroundColor: 'red' height: '80px' }}
/>
```

##### 2. Setting HTML attributes for the `<input />`
```jsx
<Input placeholder="Search..." tabIndex="-1" clearable />
```

##### 3. Setting HTML attributes for the wrapping `<div />`
```jsx
<Input placeholder="Search..." wrapper={{ role: "presentation" }} />
```

### Solution 2: `input` slot

`Input` will leverage `InputBase` and use it as a shorthand slot, the same it does for `Icon`. The structure will be:

##### Props:
```typescript
interface IInputProps extends IInputBaseProps {
  clearable?: boolean
  icon?: ItemShorthand
  input?: ItemShorthand
}
```
##### API
- preferred:
```jsx
<Input placeholder="Search..." icon="search" clearable />
````
- shorthand alternative 1 or when we need to target the `<input />`:
```jsx
<Input input={{ placeholder: "Search..." }} icon="search" clearable />
````
- shorthand alternative 2 or when we need to target the `<input />`:
```jsx
<Input input={<InputBase placeholder="Search..." />} icon="search" clearable />
````

##### Component structure
```html
<ElementType ...> // <div />
  <InputBase ... />  // <input />
  <Icon ...>  // <span />
</ElementType>
```

##### DOM structure
```html
<div class="ui-input">
  <input type="text" placeholder="Search..." value="" />
  <span class="ui-icon" aria-hidden="true"></span>
</div>
```

This will solve all the scenarios mentioned above in the following fashion:

##### 1. Styling the `<input />`
```jsx
<Input input={{ placeholder: "Search...", styles: { backgroundColor: 'red' height: '80px' } }} icon="search" clearable />
```

##### 2. Setting HTML attributes for the `<input />`
```jsx
<Input input={{ placeholder: "Search...", tabIndex: "-1" }} clearable />
```

##### 3. Setting HTML attributes for the wrapping `<div />`
```jsx
<Input placeholder="Search..." role="presentation" />
```

**Comment by brianespinosa** [comment]
In general I think a lot of this makes sense. This also addresses a similar issue to #201 as far as usability and semantics with, "What is a `<Label />` vs a `<label />`?"

I would possibly suggest that instead of `<InputBase />` we use `<InputWrapper />`. Unless that is diverging from other `<SomethingBase />` naming conventions already starting to be used in Stardust.

**Comment by bmdalex** [comment]
We decided to go with **solution 1**, implemented in #326

**Comment by layershifter** [comment]
`InputBase` represents the `input` element and is usable only with it in fact, may be `HtmlInput` will be better? `HtmlLabel` for `label` looks fine, too.

**Comment by Bugaa92** [comment]
Implemented in #326 and merged. Details:
- created an `Input` component consisting of an `<input />` element wrapped with a `<div />`;
- used `partitionHTMLProps` to decide what `props` (`<Input {...props} />`) go to the `<input />` element or wrapping `<div />`;
- used `input` slot to target the `<input />` element  and `wrapper` slot to target wrapping `<div />` when `partitionHTMLProps` does not work as expected:
```jsx
<Input {...partitionedProps} input={inputProps} wrapper={wrapperProps} />
```
e.g.:
```jsx
<Input
  placeholder='Search...'
  dir='ltr'
  tabIndex={2}
  disabled={false}
  role='presentation'
  input={{ dir: 'rtl', tabIndex: 0 }}
  wrapper={{ disabled: true, role: 'checkbox' }}
/>
```

**Comment by bmdalex** [comment]
solved and merged in #326 details:

- created an `Input` component consisting of an `<input />` element wrapped with a `<div />`;
- used `partitionHTMLProps` to decide what `props` (`<Input {...props} />`) go to the `<input />` element or wrapping `<div />`;
- used `input` slot to target the `<input />` element  and `wrapper` slot to target wrapping `<div />` when `partitionHTMLProps` does not work as expected:
```jsx
<Input {...partitionedProps} input={inputProps} wrapper={wrapperProps} />
```
e.g.:
```jsx
<Input
  placeholder='Search...'
  dir='ltr'
  tabIndex={2}
  disabled={false}
  role='presentation'
  input={{ dir: 'rtl', tabIndex: 0 }}
  wrapper={{ disabled: true, role: 'checkbox' }}
/>
```
